runtime/src/prioritization_fee.rs

```diff
@@ -0,0 +1,354 @@
+/// write account minimum fees are those greater than block minimum fee, because the minimum fee needed to land
+/// a transaction is determined by Max( min_transaction_fee, min_writable_account_fees(key), ...)
+#[derive(Debug)]
+pub struct PrioritizationFee {
+    // The minimum prioritization fee of transactions that landed in this block.
+    min_transaction_fee: u64,
+
+    // The minimum prioritization fee of each writable account in transactions in this block.
+    min_writable_account_fees: HashMap<Pubkey, u64>,
+
+    // Default to `false`, set to `true` when a block is completed, therefore the minimum fees recorded
+    // are finalized, and can be made available for use (e.g., RPC query)
+    is_finalized: bool,
+
+    // slot prioritization fee metrics
+    metrics: PrioritizationFeeMetrics,
+}
+
+impl Default for PrioritizationFee {
+    fn default() -> Self {
+        PrioritizationFee {
+            min_transaction_fee: u64::MAX,
+            min_writable_account_fees: HashMap::new(),
+            is_finalized: false,
+            metrics: PrioritizationFeeMetrics::default(),
+        }
+    }
+}
+
+impl PrioritizationFee {
+    /// Use `sanitized_tx` to update self for minimum transaction fee in the block and minimum fee for each writable account.
+    pub fn update(
+        &mut self,
+        sanitized_tx: &SanitizedTransaction,
+    ) -> Result<(), PrioritizationFeeError> {
+        let (_, update_time) = measure!(
+            {
+                let account_locks = sanitized_tx
+                    .get_account_locks()
+                    .or(Err(PrioritizationFeeError::FailGetTransactionAccountLocks))?;
+
+                let priority_details = sanitized_tx
+                    .get_transaction_priority_details()
+                    .ok_or(PrioritizationFeeError::FailGetTransactionPriorityDetails)?;
+
+                if priority_details.priority < self.min_transaction_fee {
+                    self.min_transaction_fee = priority_details.priority;
+                }
+                for write_account in account_locks.writable {
+                    self.min_writable_account_fees
+                        .entry(*write_account)
+                        .and_modify(|write_lock_fee| {
+                            *write_lock_fee =
+                                std::cmp::min(*write_lock_fee, priority_details.priority)
+                        })
+                        .or_insert(priority_details.priority);
+                }
+
+                self.metrics
+                    .increment_total_prioritization_fee(priority_details.priority);
+            },
+            "update_time",
+        );
+
+        self.metrics
+            .increment_total_update_elapsed_us(update_time.as_us());
+        Ok(())
+    }
+
+    /// Accounts that have minimum fees lesser or equal to the minimum fee in the block are redundant, they are
+    /// removed to reduce memory footprint.
```

Comment:
May be worth making a note that if this is called more than once per slot, the metrics won't be correct. It sounds like something that you could just call whenever you want. Not a big deal, since I assume right now it's only called once.

Reply:
Yeah, this function can be called many times for same slot. Practically, it is only called once at bank frozen, so retain() is invoked only once. But the metrics, as you rightly pointed out, will be wrong. Maybe change the metric reporting to something like this, with a little overhead?

Reply:
Is the "min" correct there?  If I add a few more non-trivial writable accounts and call it again, that number should increase, yes?

Reply:
on 2nd thought, it made it a private function and to be called by mark_block_completed() to enforce pruning only happens once, so these metrics are accurate.

Reply:
The best fix is one that makes it so the problem can't happen in the first place :)